doc/specs/Configuration object for profiles.md

```diff
@@ -0,0 +1,132 @@
+
+## Solution Design
+
+We will add new interfaces in the `TerminalControl` namespace, called `IControlAppearance` and `ICoreAppearance`,
+which defines how `TerminalControl` and `TerminalCore` will ask for the rendering settings they need to know about
+(such as `CursorShape`). `TerminalApp` will implement this interface through a class called `AppAppearanceConfig`.
+
+We will also have `IControlSettings` require `IControlAppearance`, and `ICoreSettings` will require `ICoreAppearance`.
+That way, the control's `settings` object can itself also be used as an object that implements both appearance interfaces. We do this so we
+do not need a separate 'regular' configuration object when we wish to switch back to the 'regular' appearance from the unfocused
+appearance - we can simply pass in the settings.
+
+On the Settings Model side, there will be a new interface called `IAppearanceConfig`, which is essentially a
+combination/mirror of the appearance interfaces described earlier. A new class, `AppearanceConfig`, will implement this
+interface and so will `Profile` itself (for the same reason as earlier - so that no new configuration object is
+needed for the regular appearance). We are choosing to have a separate interface on the settings model side to maintain
+its separation from the rest of the terminal.
+
+When we parse out the settings json file, each state-appearance will be stored in an object of the `AppearanceConfig`
+class. Later on, these values get piped over to the `AppAppearanceConfig` objects in `TerminalApp`. This is the
+similar to the way we already pipe over information such as `FontFace` and `CursorStyle` from the settings
+model to the app.
+
+### Allowed parameters
+
+For now, these states are meant to be entirely appearance-based. So, not all parameters which can be
+defined in a `Profile` can be defined in this new object (for example, we do not want parameters which
+would cause a resize in this object.) Here is the list of parameters we will allow:
+
+- Anything regarding colors: `colorScheme`, `foreground`, `background`, `cursorColor` etc
+- Anything regarding background image: `path`, `opacity`, `alignment`, `stretchMode`
+- `cursorShape`
+
+We may wish to allow further parameters in these objects in the future (like `bellStyle`?). The addition
+of further parameters can be discussed in the future and is out of scope for this spec.
+
+### Inheritance
+
+In the case that not all of the allowed parameters are defined in an appearance object, we will obtain the
+values for those parameters in the following matter:
+
+If the profile defines an `unfocusedConfig`, any parameters not explicitly defined within it will adopt
+the values from the profile itself. If the profile does not define an `unfocusedConfig`, then the global/default `unfocusedConfig` is used
+for this profile.
+
+Thus, if a user wishes for the unfocused state to look the same as the focused state for a particular profile,

```

Comment:
Okay, so just to be sure, the ordering is 

```py
profile.unfocusedState or
profiles.defaults.unfocusedState or
profile.appearance or
profile.defaults.appearance
```

_(where `profile.appearance` is just the appearance of the profile itself)_

and we just return whatever the first non-null one is? And if the `profile.unfocusedState` is `{}`, we'll just use `profile.appearance`?

Reply:
Yes, though there is also the question of what we do about the _undefined_ parameters within the unfocused config object, so I prefer to think of it as:

`profile.unfocusedConfig` defined: undefined params taken from `profile` (so the case where `profile.unfocusedConfig` is `{}` falls in here)
`profile.unfocusedConfig` not defined, `profile.defaults.unfocusedConfig` defined: undefined params are default values
both not defined: do nothing (so it just keeps the profile's appearance when unfocused)

Reply:
OKay I think case 2 is _weird_, but it's well defined at least. I guess I'd expect undefined params in that case to use the profile ones still, but I understand how (from a layering perspective), that's just impossible. I think it'd be weird to throw an unfocused config in the defaults...


okay though what if I wanted all my profiles to be _gray_ when they're unfocused. I'd throw a `unfocusedConfig: { scheme: gray }` in the defaults, and be happy. But then the _rest_ of the unfocused config is coming from the default settings, not the profile's settings.

Like yea that makes sense, it's just _weird_

Reply:
nit: I would add a small comment somewhere explaining why we're doing this "all-or-nothing" approach as opposed to the "per-setting" approach.

Reply:
> Like yea that makes sense, it's just _weird_

Yeah but its a clean way given we _don't_ want to define our own default `unfocusedConfig` and we don't want to mess up when no `unfocusedConfig` is defined anywhere in the file

> nit: I would add a small comment somewhere explaining why we're doing this "all-or-nothing" approach as opposed to the "per-setting" approach.

Sounds good, will do